Mathlib/Lean/Meta/DiscrTree.lean

```diff
@@ -53,19 +53,25 @@ partial def getSubexpressionMatches (d : DiscrTree α s) (e : Expr) : MetaM (Arr
 
 variable {m : Type → Type} [Monad m]
 
-/-- Apply a monadic function to the array of values at each node in a `DiscrTree`. -/
-partial def Trie.mapArraysM (t : DiscrTree.Trie α s) (f : Array α → m (Array β)) :
-    m (DiscrTree.Trie β s) := do
-  match t with
-  | .node vs children =>
-    return .node (← f vs) (← children.mapM fun (k, t') => do pure (k, ← t'.mapArraysM f))
 
-/-- Apply a monadic function to the array of values at each node in a `DiscrTree`. -/
-def mapArraysM (d : DiscrTree α s) (f : Array α → m (Array β)) : m (DiscrTree β s) := do
-  pure { root := ← d.root.mapM (fun t => t.mapArraysM f) }
+/-- A better representaton for lists-of-quadruples -/
+private inductive Ctxt (α β γ δ : Type)
+  | empty : Ctxt α β γ δ
+  | ctxt : α → β → γ → δ → Ctxt α β γ δ → Ctxt α β γ δ
```

Comment:
Can you explain what "better" means here? I assume it's some combination of:
* Smaller memory footprint
* This being a bottleneck somehow
* Tail recursion not working for a nested recursor

Reply:
Only smaller memory footprint, one could use `List (a × b × c d)` here as well, and it would still be tail recursive. Maybe a bit premature optimization, but it doesn't hurt.

Reply:
```suggestion
/-- A representation of lists-of-quadruples with a smaller memory footprint. -/
private inductive Ctxt (α β γ δ : Type)
  | empty : Ctxt α β γ δ
  | ctxt : α → β → γ → δ → Ctxt α β γ δ → Ctxt α β γ δ
```